[SPARK-8992] [SQL] Add pivot to dataframe api

aray opened the pull request:
This adds a pivot method to the dataframe api.

Following the lead of cube and rollup this adds a Pivot operator that is translated into an Aggregate by the analyzer.

Currently the syntax is like:
~~courseSales.pivot(Seq($"year"), $"course", Seq("dotNET", "Java"), sum($"earnings"))~~

~~Would we be interested in the following syntax also/alternatively? and~~

```
courseSales.groupBy($"year").pivot($"course", "dotNET", "Java").agg(sum($"earnings"))
//or
courseSales.groupBy($"year").pivot($"course").agg(sum($"earnings"))
```

Later we can add it to `SQLParser`, but as Hive doesn't support it we cant add it there, right?

~~Also what would be the suggested Java friendly method signature for this?~~

aray committed:
Add pivot to dataframe api

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

rxin commented:
Jenkins, ok to test.

aray commented:
@rxin it looks like Jenkins forgot about building this. Can you help trigger the build again?

SparkQA commented:
[Test build #1319 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1319/consoleFull) for   PR 7841 at commit [`599e9e0`](https://github.com/apache/spark/commit/599e9e0b9bd46be798da1274e9ba9839151b2aaf).

SparkQA commented:
[Test build #1319 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1319/console) for   PR 7841 at commit [`599e9e0`](https://github.com/apache/spark/commit/599e9e0b9bd46be798da1274e9ba9839151b2aaf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Pivot(`

aray committed:
fix unit test answer

rxin commented:
@aray FYI this didn't make it into the 1.5 release (was submitted too close to the feature freeze deadline), but we will try to include it in Spark 1.6.

aray committed:
Merge branch 'master' of https://github.com/apache/spark into sql-pivot

Conflicts:
	sql/core/src/test/scala/org/apache/spark/sql/TestData.scala

JoshRosen commented:
@rxin, do you want to revisit this now for 1.6?

aray committed:
Support for pivot as operation on GroupedData like:
courseSales.groupBy($"year").pivot($"course", "dotNET", "Java").agg(sum($"earnings"))
Also, fixed master merge.

aray committed:
Merge branch 'master' of https://github.com/apache/spark into sql-pivot

Conflicts:
	sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

aray committed:
fix some style issues and remove commented import

aray commented:
@rxin and @JoshRosen, this is ready for review now.

rxin commented:
@aray Thanks a lot for updating this. To help api design, can you take a look at other frameworks and see what their signatures look like?

aray commented:
@rxin here is my summary of other frameworks API's

I'm going to use an example dataset form the pandas doc for all the examples (as df)

|A|B|C|D|
|---|---|---|---|
|foo|one|small|1|
|foo|one|large|2|
|foo|one|large|2|
|foo|two|small|3|
|foo|two|small|3|
|bar|one|large|4|
|bar|one|small|5|
|bar|two|small|6|
|bar|two|large|7|

This API
--------

```scala
scala> df.groupBy("A", "B").pivot("C", "small", "large").sum("D").show
+---+---+-----+-----+
|  A|  B|small|large|
+---+---+-----+-----+
|foo|two|    6| null|
|bar|two|    6|    7|
|foo|one|    1|    4|
|bar|one|    5|    4|
+---+---+-----+-----+

scala> df.groupBy("A", "B").pivot("C", "small", "large").agg(sum("D"), avg("D")).show
+---+---+------------+------------+------------+------------+
|  A|  B|small sum(D)|small avg(D)|large sum(D)|large avg(D)|
+---+---+------------+------------+------------+------------+
|foo|two|           6|         3.0|        null|        null|
|bar|two|           6|         6.0|           7|         7.0|
|foo|one|           1|         1.0|           4|         2.0|
|bar|one|           5|         5.0|           4|         4.0|
+---+---+------------+------------+------------+------------+

scala> df.pivot(Seq($"A", $"B"), $"C", Seq("small", "large"), sum($"D")).show
+---+---+-----+-----+
|  A|  B|small|large|
+---+---+-----+-----+
|foo|two|    6| null|
|bar|two|    6|    7|
|foo|one|    1|    4|
|bar|one|    5|    4|
+---+---+-----+-----+
```

We require a list of values for the pivot column as we are required to know the output columns of the operator ahead of time. Pandas and reshape2 do not require this but the comparable SQL operators do. We also allow multiple aggregations which not all implementations allow.

pandas
------

The comparable metod for pandas is `pivot_table(data, values=None, index=None, columns=None, aggfunc='mean', fill_value=None, margins=False, dropna=True)`

Example

```python
>>> pivot_table(df, values='D', index=['A', 'B'], columns=['C'], aggfunc=np.sum)
          small  large
foo  one  1      4
     two  6      NaN
bar  one  5      4
     two  6      7
```

Pandas also allows multiple aggregations:

```python
>>> pivot_table(df, values='D', index=['A', 'B'], columns=['C'], aggfunc=[np.sum, np.average])
          sum       average      
C       large small   large small
A   B                            
bar one     4     5       4     5
    two     7     6       7     6
foo one     4     1       2     1
    two   NaN     6     NaN     3
```

References

- http://pandas.pydata.org/pandas-docs/stable/reshaping.html
- http://pandas.pydata.org/pandas-docs/stable/generated/pandas.pivot_table.html

See also: `pivot`, `stack`, `unstack`.

reshape2 (R)
------------
The comparable method for reshape2 is `dcast(data, formula, fun.aggregate = NULL, ..., margins = NULL, subset = NULL, fill = NULL, drop = TRUE, value.var = guess_value(data))`

```r
> dcast(df, A + B ~ C, sum)
Using D as value column: use value.var to override.
    A   B large small
1 bar one     4     5
2 bar two     7     6
3 foo one     4     1
4 foo two     0     6
```

Note that by default cast fills with the value from applying fun.aggregate to 0 length vector

References

- https://cran.r-project.org/web/packages/reshape2/reshape2.pdf
- http://seananderson.ca/2013/10/19/reshape.html
- http://www.inside-r.org/packages/cran/reshape2/docs/cast

See also: `melt`.

MS SQL Server
----------

```sql
SELECT *
FROM df
pivot (sum(D) for C in ([small], [large])) p
```

http://sqlfiddle.com/#!3/cf887/3/0

References

- http://sqlhints.com/2014/03/10/pivot-and-unpivot-in-sql-server/


Oracle 11g
----------

```sql
SELECT *
FROM df
pivot (sum(D) for C in ('small', 'large')) p
```
http://sqlfiddle.com/#!4/29bc5/3/0

Oracle also allows multiple aggregations and with similar output to this api

```sql
SELECT *
FROM df
pivot (sum(D) as sum, avg(D) as avg for C in ('small', 'large')) p
```
http://sqlfiddle.com/#!4/29bc5/5/0

References

- http://www.oracle.com/technetwork/articles/sql/11g-pivot-097235.html
- http://docs.oracle.com/cd/B28359_01/server.111/b28286/statements_10002.htm#CHDCEJJE
- http://www.techonthenet.com/oracle/pivot.php

----------

Let me know if I can do anything else to help this along. Also would you mind adding me to the jenkins whitelist so I can test it?

marmbrus commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44249 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44249/consoleFull)** for PR 7841 at commit [`f2827ea`](https://github.com/apache/spark/commit/f2827ea82c8f8f7f44552d820c12a693fe25aaa9).

SparkQA commented:
**[Test build #44249 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44249/consoleFull)** for PR 7841 at commit [`f2827ea`](https://github.com/apache/spark/commit/f2827ea82c8f8f7f44552d820c12a693fe25aaa9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Pivot(`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44249/
Test PASSed.

rxin commented:
I like your 2nd interface more (group by and then pivot), since it is easier to get that working for both Java and Scala. We can implement a simpler interface for Python/R that's closer to existing frameworks.

How hard would it be to not require the values?

aray commented:
@rxin, Not requiring the values would necessitate doing a separate query for the distinct values of the column before the pivot query. It looks like at least some DF operations (eg, drop) would need the result so even if we made Pivot.output lazy we would be running an unexpected job.

If a user really didn't want to specify the values, they can explicitly do the query:

```scala
    df.groupBy("A", "B").pivot("C", df.select("C").distinct.collect.map(_.getString(0)): _*).sum("D")
```

Needing to know the output columns of an operator for analysis/planning is probably why the other SQL implementations require the values also (technically Oracle supports omitting it but only in XML mode where you essentially just get one column).

aray committed:
Fix new merge conflict

Merge branch 'master' of https://github.com/apache/spark into sql-pivot

Conflicts:
	sql/core/src/main/scala/org/apache/spark/sql/GroupedData.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44643 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44643/consoleFull)** for PR 7841 at commit [`403f966`](https://github.com/apache/spark/commit/403f966de21e761fe01127949af697d00de91cc4).

SparkQA commented:
**[Test build #44643 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44643/consoleFull)** for PR 7841 at commit [`403f966`](https://github.com/apache/spark/commit/403f966de21e761fe01127949af697d00de91cc4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Pivot(`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44643/
Test PASSed.

rxin commented:
@aray sorry was away for spark summit - back now and will get to this today.

aray committed:
Fix merge conflicts again

Merge branch 'master' of https://github.com/apache/spark into sql-pivot

Conflicts:
	sql/core/src/main/scala/org/apache/spark/sql/GroupedData.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45316 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45316/consoleFull)** for PR 7841 at commit [`d8e473c`](https://github.com/apache/spark/commit/d8e473c64f3796a88d809a0e061329d4e359c6bf).

SparkQA commented:
**[Test build #45316 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45316/consoleFull)** for PR 7841 at commit [`d8e473c`](https://github.com/apache/spark/commit/d8e473c64f3796a88d809a0e061329d4e359c6bf).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Pivot(`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

aray committed:
Fix scala style issue from merge

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45366 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45366/consoleFull)** for PR 7841 at commit [`2417548`](https://github.com/apache/spark/commit/2417548ee88369fc45547346506c8fa4feb2010e).

SparkQA commented:
**[Test build #45366 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45366/consoleFull)** for PR 7841 at commit [`2417548`](https://github.com/apache/spark/commit/2417548ee88369fc45547346506c8fa4feb2010e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Pivot(`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

rxin commented:
@aray I talked to a few more people about this. Most like the 2nd API more (groupBy.pivot.agg).

I think it'd also be better to remove the requirement to specify values, e.g. just take in a column without the values. So it looks like
```scala
courseSales.groupBy($"year").pivot($"course").agg(sum($"earnings"))
```

Can you update the pull request? Thanks.

rxin commented:
BTW we can also later add a variant that allows users to specify values directly, in order to avoid materializing the intermediate data.

aray committed:
Update pivot to make values optional, call .distinct() on column if not provided. Add unit tests for this scenario.

aray commented:
@rxin Updated, the values are now optional.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45432 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45432/consoleFull)** for PR 7841 at commit [`1af796d`](https://github.com/apache/spark/commit/1af796d45f58fccc46f5c49ec200f2bbc97e8d9d).

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala:
```diff
@@ -989,6 +989,41 @@ class DataFrame private[sql](
+   * @group dfops
+   * @since 1.6.0
+   */
+  def pivot(
```
can we remove this?

aray replied:
sure

SparkQA commented:
**[Test build #45432 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45432/consoleFull)** for PR 7841 at commit [`1af796d`](https://github.com/apache/spark/commit/1af796d45f58fccc46f5c49ec200f2bbc97e8d9d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Pivot(`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45432/
Test PASSed.

aray committed:
Remove DataFrame.pivot monster method

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45486 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45486/consoleFull)** for PR 7841 at commit [`6e3b133`](https://github.com/apache/spark/commit/6e3b1337f9da8a5c15b1bef458f25f11fe94dbcd).

SparkQA commented:
**[Test build #45486 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45486/consoleFull)** for PR 7841 at commit [`6e3b133`](https://github.com/apache/spark/commit/6e3b1337f9da8a5c15b1bef458f25f11fe94dbcd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Pivot(`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45486/
Test PASSed.

aray committed:
Fix merge conflicts

Merge branch 'master' of https://github.com/apache/spark into sql-pivot

Conflicts:
	sql/core/src/main/scala/org/apache/spark/sql/GroupedData.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45565 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45565/consoleFull)** for PR 7841 at commit [`04d643c`](https://github.com/apache/spark/commit/04d643cd7dc7f3cb9b46b385b2940be8ab9eb1f6).

SparkQA commented:
**[Test build #45565 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45565/consoleFull)** for PR 7841 at commit [`04d643c`](https://github.com/apache/spark/commit/04d643cd7dc7f3cb9b46b385b2940be8ab9eb1f6).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45565/
Test FAILed.

rxin commented:
To give a status update -- I'm hoping to merge this this week.

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -247,6 +252,38 @@ class Analyzer(
+              case u: UnaryExpression if u.isInstanceOf[AggregateExpression] =>
+                u.withNewChildren(Seq(
+                  If(EqualTo(pivotColumn, Literal(value)), u.child, Literal(null))
+                ))
```
@aray We have switched to our new `AggregateExpression` and `AggregateFunction` interfaces. With the new interface, we always use `AggregateExpression` to wrap an `AggregateFunction`. So, I think the matching condition can be changed to match `AggregateFunction` instead of matching `AggregateExoression`.

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -247,6 +252,38 @@ class Analyzer(
+              throw new AnalysisException(
+                s"Unary aggregate expression required for pivot, found '$aggregate'")
+            }
+            val name = if (singleAgg) value else value + " " + aggregate.prettyString
```
Seems it is always helpful to show the aggregate function(s) that we use for pivot?

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -248,6 +253,38 @@ class Analyzer(
+      case Pivot(groupByExprs, pivotColumn, pivotValues, aggregates, child) =>
+        val singleAgg = aggregates.size == 1
+        val pivotAggregates: Seq[NamedExpression] = pivotValues.flatMap{ value =>
+          aggregates.map{ aggregate =>
```
`aggregates.map { aggregate =>`?

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -248,6 +253,38 @@ class Analyzer(
+            val filteredAggregate = aggregate.transformDown{
+              case u: UnaryExpression if u.isInstanceOf[AggregateExpression] =>
+                u.withNewChildren(Seq(
+                  If(EqualTo(pivotColumn, Literal(value)), u.child, Literal(null))
```
I guess the underlying assumption of this line is that aggregate function should ignore `null` input values, right? (Basically, `null` should have no impact on the result of this aggregate function.) If so, can we add a comment at here?

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -248,6 +253,38 @@ class Analyzer(
+            if (filteredAggregate.fastEquals(aggregate)) {
+              throw new AnalysisException(
+                s"Unary aggregate expression required for pivot, found '$aggregate'")
+            }
```
It is not super obvious why we may hit here. Can we add an example?

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -248,6 +253,38 @@ class Analyzer(
+            }
+            if (filteredAggregate.fastEquals(aggregate)) {
+              throw new AnalysisException(
+                s"Unary aggregate expression required for pivot, found '$aggregate'")
```
Maybe we should explicitly say that if we hit this branch, the function is not an aggregate function (the case above match non-unary aggregate functions)?

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -997,6 +1034,7 @@ class Analyzer(
       case p if !p.resolved => p // Skip unresolved nodes.
       case p: Project => p
       case f: Filter => f
+      case p: Pivot => p
```
Why?

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala:
```diff
@@ -385,6 +385,20 @@ case class Rollup(
+  override def output: Seq[Attribute] = groupByExprs.map(_.toAttribute) ++ aggregates match {
+    case aggregate :: Nil => pivotValues.map(AttributeReference(_, aggregate.dataType)())
+    case _ => pivotValues.flatMap{ value =>
+      aggregates.map(agg => AttributeReference(value + " " + agg.prettyString, agg.dataType)())
```
Same question. Should we always add aggregate function in the name?

yhuai replied:
btw, instead of using a whitespace, will `_` be better?

aray replied:
In other comparable implementations (see above) the aggregate function name is not included unless multiple are used. I will change the separator to `_`.

yhuai replied:
ah i see. Thank you for the explanation!

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/GroupedData.scala:
```diff
@@ -273,6 +280,60 @@ class GroupedData protected[sql](
+    *   df.groupBy($"year").pivot($"course", "dotNET", "Java").agg(sum($"earnings"))
+    *   // Or without specifying column values
+    *   df.groupBy($"year").pivot($"course").agg(sum($"earnings"))
+    * }}}
```
How about we let users know that if no pivot values are provided, we will launch a job to find all distinct values of the pivot column?

yhuai replied:
Another question I have is that, what will happen if we have too many distinct values? I am wondering if we should always ask users to put pivot values?

aray replied:
There is a note below in the description of the values parameter
>If values are not provided the method with do an immediate call to .distinct() on the pivot column.
Do we need to duplicate/move that note to the method description?

Right now if there is too many distinct values we probably get OOM. Obvious solve is to have some configurable maximum above which we give an error. Should I try to add that now?

yhuai replied:
Yeah. Let's add that. In our error message, we can ask users to change the value of that conf (btw, let's make sure the default value of that conf is large enough for common use cases).

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/GroupedData.scala:
```diff
@@ -273,6 +280,60 @@ class GroupedData protected[sql](
+        values
+      } else {
+        // Get the distinct values of the column and sort them so its consistent
+        df.select(pivotColumn.cast(StringType))
```
Do we need this cast?

yhuai replied:
If we cast values to strings, in `EqualTo(pivotColumn, Literal(value)` (in analyzer), we will create string literals. My concern is that turning those values to strings may make `EqualTo(pivotColumn, Literal(value)` return wrong results.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/GroupedData.scala:
```diff
@@ -273,6 +280,60 @@ class GroupedData protected[sql](
+    * @since 1.6.0
+    */
+  @scala.annotation.varargs
+  def pivot(pivotColumn: Column, values: String*): GroupedData = groupType match {
```
For values, should we change the type to `values: Column*`? Every value is a literal created by `lit` function.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/GroupedData.scala:
```diff
@@ -273,6 +280,60 @@ class GroupedData protected[sql](
+  def pivot(pivotColumn: String, values: String*): GroupedData = {
+    val resolvedPivotColumn = Column(df.resolve(pivotColumn))
+    pivot(resolvedPivotColumn, values: _*)
+  }
```
For the first version, maybe we can just have the API using `Column` as the argument type? (I am thinking about the type of values. I am not sure String is the right type).

yhuai commented:
@aray This is very cool! Here are a few things I'd like to discuss.
1. Should we always ask users to provide pivot values?
2. For the type of pivot values, should we just use String? Or, we should use Literal Column?
3. For the column name of a pivot aggregation column, should we always show the aggregate function in a column name? (right now, we only show function name if we have more than 1 aggregations)

aray committed:
Address comments in Analyzer

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45645 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45645/consoleFull)** for PR 7841 at commit [`88dd513`](https://github.com/apache/spark/commit/88dd513d69c7a94274cb37e76aa854ddceb977cb).

SparkQA commented:
**[Test build #45645 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45645/consoleFull)** for PR 7841 at commit [`88dd513`](https://github.com/apache/spark/commit/88dd513d69c7a94274cb37e76aa854ddceb977cb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class SlidingRDDPartition[T](val idx: Int, val prev: Partition, val tail: Seq[T], val offset: Int)`\n  * `class SlidingRDD[T: ClassTag](@transient val parent: RDD[T], val windowSize: Int, val step: Int)`\n  * `case class Pivot(`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45645/
Test PASSed.

aray committed:
Address remaining comments
- Use Literal's for the pivot column values instead of strings.
- Change seperator when using multiple aggregates to `_` instead of space.
- Some additional unit testing

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45659 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45659/consoleFull)** for PR 7841 at commit [`12a8270`](https://github.com/apache/spark/commit/12a8270b7592c0953464e2353fafbfbbeda92f3a).

aray commented:
@yhuai RE your questions (3 was already addressed above):
>1. Should we always ask users to provide pivot values?
The argument for not requiring values I think is convenience and also similarity to other non-sql tools mentioned above. The negative is performance, but since we give them the option to specify I don't think that is a problem.
>2. For the type of pivot values, should we just use String? Or, we should use Literal Column?
I initially used strings as the type since that is the common usage scenario. But I agree that using Literal's is the better solution and will avoid casts which could hurt performance. For convenience I kept the second method (changed to `pivot(pivotColumn: String, values: Any*)`) which I think will be the prefered way to use pivot since it is more succinct.

I really appreciate the review. Let me know if I can do anything else to help!

yhuai commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -248,6 +253,43 @@ class Analyzer(
+              case Last(expr, _) =>
+                Last(ifExpr(expr), Literal(true))
+              case a: AggregateFunction =>
+                a.withNewChildren(a.children.map(ifExpr))
```
Seems we still need to check the number of children and make sure we have a single child?

aray replied:
It should now work fine with aggregate functions that have multiple children as long as they ignore updates when all values are null. For example `Corr` should work since it only updates its aggregation buffer if both its arguments are non null.

yhuai replied:
oh, yes. You are right.

SparkQA commented:
**[Test build #45659 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45659/consoleFull)** for PR 7841 at commit [`12a8270`](https://github.com/apache/spark/commit/12a8270b7592c0953464e2353fafbfbbeda92f3a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Pivot(`\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45659/
Test FAILed.

aray committed:
Add configurable maximum number of pivot values when none are given to prevent unintended OOM errors.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

aray commented:
@yhuai I think this addresses everything we discussed, let me know if I missed anything or if there is anything else I can do. Again, thanks for the code review.

SparkQA commented:
**[Test build #45673 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45673/consoleFull)** for PR 7841 at commit [`676f1ac`](https://github.com/apache/spark/commit/676f1accde4ce7cbbb7f274315dc3ab6e679d3db).

yhuai commented:
LGTM pending jenkins.

SparkQA commented:
**[Test build #45673 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45673/consoleFull)** for PR 7841 at commit [`676f1ac`](https://github.com/apache/spark/commit/676f1accde4ce7cbbb7f274315dc3ab6e679d3db).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `case class Pivot(`\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45673/
Test PASSed.

yhuai commented:
Thanks! Merging to master and branch 1.6.

rxin commented:
@aray do you want to submit a pull request for python api too?

aray commented:
@rxin sure I'll put together a PR for the python API tonight

rxin commented:
@aray this pull request was highlighted in http://www.slideshare.net/databricks/deep-dive-into-catalyst-apache-spark-20s-optimizer

pushpam002 commented:
thank [you ](https://www.trickcode.in/)
